worker/src/lib.rs

```diff
@@ -0,0 +1,615 @@
+                lv2_sys::LV2_Worker_Status_LV2_WORKER_ERR_NO_SPACE => Err(WorkerError::NoSpace),
+                _ => Err(WorkerError::Unknown),
+            }
+        }
+    }
+}
+
+impl<P: Worker> From<Schedule<'_>> for ScheduleHandler<P> {
+    fn from(schedule_feature: Schedule) -> Self {
+        Self {
+            schedule_handle: schedule_feature.internal.handle,
+            schedule_work: schedule_feature.internal.schedule_work,
+            phantom: PhantomData::<P>,
+        }
+    }
+}
+
+/// Handler available inside worker function to send response to `run()` context.
+///
+/// ResponseHandler need the current Worker trait implementor as generic parameter, because it use
+/// the `ResponseData` associated type to know the datatype to send to the `worker_response`
+/// method.
+pub struct ResponseHandler<P: Worker> {
+    /// function provided by the host to send response to `run()`
+    response_function: lv2_sys::LV2_Worker_Respond_Function,
+    /// Response handler provided by the host, must be passed to the host provided
+    /// response_function.
+    respond_handle: lv2_sys::LV2_Worker_Respond_Handle,
+    phantom: PhantomData<P>,
+}
+
+impl<P: Worker> ResponseHandler<P> {
+    /// Allow to give response to the `run` context.
+    ///
+    /// This method allow to give a response to the `run` context. After calling this method, the host will call
+    /// `worker_response` and give the passed data.
+    ///
+    /// **Notes about the passed data:** The buffer used to pass data is managed by the host. That
+    /// mean the size is unknown and may be limited. So if you need to pass huge amount of data,
+    /// it's preferable to use another way, for example a sync::mpsc channel.
+    pub fn respond(&self, response_data: P::ResponseData) -> Result<(), WorkerError>
+    where
+        P::WorkData: 'static + Send,
+    {
+        unsafe {
+            let response_data = mem::ManuallyDrop::new(response_data);
+            let size = mem::size_of_val(&response_data) as u32;
+            let ptr = &response_data as *const _ as *const c_void;
+            let response_function = if let Some(response_function) = self.response_function {
+                response_function
+            } else {
+                return Err(WorkerError::Unknown);
+            };
+            match (response_function)(self.respond_handle, size, ptr) {
+                lv2_sys::LV2_Worker_Status_LV2_WORKER_SUCCESS => Ok(()),
+                lv2_sys::LV2_Worker_Status_LV2_WORKER_ERR_UNKNOWN => Err(WorkerError::Unknown),
+                lv2_sys::LV2_Worker_Status_LV2_WORKER_ERR_NO_SPACE => Err(WorkerError::NoSpace),
+                _ => Err(WorkerError::Unknown),
+            }
+        }
+    }
+}
+
+/// Errors potentially generated by [`Worker`](trait.Worker.html) methods
+pub enum WorkerError {
+    /// Unknown or general error
+    Unknown,
+    /// Failure due to a lack of space
+    NoSpace,
+}
```

Comment:
sorry, this one is confuse to me. you mean have different enum between Work and response, or you mean using a different error when some callback function didn't exist ?

Reply:
I mean that you should add more variants to this enum, since there more types of errors that may happen when someone uses your code.

Reply:
you mean something like `NoCallback` ?

Reply:
For example, yes.

Reply:
i did big change here in my local work. The main reason is because `schedule_work` and `respond` was causing memory leak when failing to send data containing pointers. So i used the same strategy in `std::mpsc::sender`, return the data in the error.
I added "NoCallback" variant for my new `ScheduleError` and `RespondError`.